java/arcs/core/data/expression/Evaluator.kt

```diff
@@ -93,17 +95,43 @@ class ExpressionEvaluator(
     }
 
     override fun visit(expr: Expression.NewExpression): Any {
-        val newScope = scopeCreator(expr.schemaName.firstOrNull() ?: "")
-        expr.fields.forEach { (fieldName, fieldExpr) ->
-            newScope.set(fieldName, fieldExpr.accept(this))
+        val initScope = scopeCreator(expr.schemaName.firstOrNull() ?: "")
+        return expr.fields.fold(initScope) { scope, (fieldName, fieldExpr) ->
```

Comment:
This will create O(N) new immutable and mutable maps for N being the number of fields. Can we do better and create the map once here and turn it into the scope?

Reply:
Added nested immutable scopes